Support generic templates

rohanjr opened the pull request:
This change supports compilation of generic templates in Haskell with DAML extensions enabled. Requesting review while I test in DAML proper.

rohanjr committed:
Parse (single) template constraint and pass to mkTemplateDecls where it is ignored

rohanjr committed:
Pass template constraints through to TInstance class declaration

rohanjr committed:
Pass type vars into TInstance class decl

rohanjr committed:
Only generate base class instance if the template is not generic

rohanjr committed:
Include TInstance constraint in Choice and Key instances

rohanjr committed:
Use type parameters in template and choice data decls

rohanjr committed:
Pass type parameters to uses of template type

rohanjr committed:
Use template/choice data type parameters in choice method signatures

rohanjr committed:
Don't apply Archive data type to type parameters

rohanjr committed:
Abstract template/choice type construction functions, prepare for tyvars in instances

rohanjr committed:
Use type parameters for template/choice data types in instance decls

rohanjr committed:
Pass in type variables for TInstance context

rohanjr committed:
Parse template instance and don't generate anything

rohanjr committed:
Make instance decl for template instance

rohanjr committed:
Produce new type decl for template instance

rohanjr committed:
Replace custom template constraints with context used in rest of parser

rohanjr committed:
Revert "Replace custom template constraints with context used in rest of parser"

This reverts commit 5dc09d9d072f5be9d7b887e6b481994110aa7885.

rohanjr committed:
Allow parsing multiple constraints in templates

hurryabit commented on compiler/parser/Parser.y:
```diff
@@ -1157,8 +1157,38 @@ topdecl :: { LHsDecl GhcPs }
-  : 'template' qtycon arecord_with 'where' template_body
+  : 'template' template_header arecord_with 'where' template_body
                                                  {% mkTemplateDecls $2 $3 $5 }
+  | 'template' 'instance' qtycon '=' qtycon qtycons
```
This rule is not general enough. We want to allow things like
```haskell
template instance ProposalXorY = Proposal (Either X Y)
```
as well. That is, instead of `qtycons` we need a list of types.

rohanjr replied:
Good point, I will change.

hurryabit commented on compiler/parser/Parser.y:
```diff
@@ -1157,8 +1157,38 @@ topdecl :: { LHsDecl GhcPs }
+  | 'template' 'instance' qtycon '=' qtycon qtycons
+                                                 {% mkTemplateInstance $3 $5 (unLoc $6) }
+
+template_header :: { Located TemplateHeader }
```
Can we do this a la `tycl_hdr`? That is
```haskell
template_header : context '=>' qtycon tyvars | qtycon tyvars
```

rohanjr replied:
See the NOTE below. I tried that in this commit https://github.com/digital-asset/ghc/pull/32/commits/5dc09d9d072f5be9d7b887e6b481994110aa7885, but it caused failure to parse non-generic templates. It could be recoverable though, with some rejigging of the parse rules perhaps. It would be nice to parse a context directly, instead of parsing the constraints and converting them to an LHsContext later.

hurryabit replied:
Let's merge it as is then but try to improve in a subsequent PR. I'm convinced it is possible to reuse GHC's rules for this.

hurryabit commented on compiler/parser/Parser.y:
```diff
@@ -1157,8 +1157,38 @@ topdecl :: { LHsDecl GhcPs }
+                                                 {% mkTemplateInstance $3 $5 (unLoc $6) }
+
+template_header :: { Located TemplateHeader }
+  : qtycon                                       { sL1 $1 $ TemplateHeader [] $1 [] }
```
This case is a special case of the next one since `varids0` allows for the empty list.

rohanjr replied:
Good point, will fix.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2261,9 +2275,29 @@ mkUpdate = mkAppTy (mkQualType "Update")
+mkAppTyArgs :: LHsType GhcPs -> [Located RdrName] -> LHsType GhcPs
+mkAppTyArgs tyCon tyVars = foldl' mkAppTy tyCon $ map rdrNameToType tyVars
+
+mkTemplateType :: String -> [Located RdrName] -> LHsType GhcPs
```
```suggestion
mkTemplateType :: TemplateHeader -> LHsType GhcPs
```
and pass `TemplateHeader` around instead of its components.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2261,9 +2275,29 @@ mkUpdate = mkAppTy (mkQualType "Update")
+mkChoiceType :: Located RdrName -> [Located RdrName] -> LHsType GhcPs
+mkChoiceType choiceName tyVars =
+  case occNameString $ rdrNameOcc $ unLoc choiceName of
+    "Archive" -> rdrNameToType choiceName
```
Can we only do this if `Archive` is the one from `DA.Internal.Template`? What would otherwise happen if I call a user written choice `Archive`? (This will lead to disaster later anyway, but I'm afraid of very bad error messages.)

rohanjr replied:
Hmm I'm not sure how this would impact the error message. This just decides whether to drop the type parameters in occurrences of the choice data type. It should give the same error of ambiguous data types called `Archive`. I don't mind making the check more specific, but we'll have to this code anyway if we choose your approach of dropping unused type parameters.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2261,9 +2275,29 @@ mkUpdate = mkAppTy (mkQualType "Update")
 partiesType :: LHsType GhcPs
 partiesType = noLoc $ HsListTy noExt $ mkQualType "Party"
 
+mkAppTyArgs :: LHsType GhcPs -> [Located RdrName] -> LHsType GhcPs
```
Does the GHC API not provide a function for this?

rohanjr replied:
Ah yeah, found some stuff like `mkHsAppTy[s]`. I'll refactor

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2261,9 +2275,29 @@ mkUpdate = mkAppTy (mkQualType "Update")
 
+-- Type variable conversions
+
+rdrNameToTyVar :: Located RdrName -> LHsTyVarBndr GhcPs
```
Does the GHC API provide a function for this?

rohanjr replied:
Don't see anything

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2355,15 +2389,15 @@ funBind loc tag mg =
 
-mkTemplateClassInstanceSigs :: String -> Maybe (LHsType GhcPs) -> [LSig GhcPs]
-mkTemplateClassInstanceSigs templateName mbKeyType =
+mkTemplateClassInstanceSigs :: String -> [Located RdrName] -> Maybe (LHsType GhcPs) -> [LSig GhcPs]
```
```suggestion
mkTemplateClassInstanceSigs :: Located RdrName -> [Located RdrName] -> Maybe (LHsType GhcPs) -> [LSig GhcPs]
```

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2381,19 +2415,25 @@ mkTemplateClassInstanceSigs templateName mbKeyType =
-    templateType = mkUnqualType templateName
-    contractId = mkContractId templateType
-    hasKeyType = mkQualType "HasKey" `mkAppTy` templateType
+    templateType = mkTemplateType templateName tyVars
```
Instead of always applying `mbParenTy` to `templateType`, we should do it here:
```suggestion
    templateType = (if null tyVars then id else mkParenTy) $ mkTemplateType templateName tyVars
```

hurryabit replied:
Actually, this happens in other places as well. Should be add the parenthesizing into `mkTemplateType`?

rohanjr replied:
I was trying to avoid unnecessary parentheses, which you can only determine at the occurrences. I could add add parens in `mkTemplateType`, but you'd end up with lots of extra (harmless) parens.

hurryabit replied:
We could add a flag to `templateType` which determines whether we want to add parentheses when they are necessary.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2585,23 +2627,40 @@ mkTemplateClassInstanceMethods conName ValidTemplateBody{..} =
                           (mkUnqualVar $ mkVarOcc "cid"))
                         (mkQualVar $ mkDataOcc "Archive")
 
+templateConstraintsToContext :: [Located TemplateConstraint] -> LHsContext GhcPs
```
Hopefully we get rid of this if we can change the parse rule `template_hdr` as I suggested.

rohanjr replied:
Hopefully - can you see how to salvage https://github.com/digital-asset/ghc/pull/32/commits/5dc09d9d072f5be9d7b887e6b481994110aa7885 from a glance?

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2614,58 +2673,69 @@ mkTemplateInstanceMethods templateName =
 
+withInstanceContext :: String -> [Located RdrName] -> LHsType GhcPs -> HsType GhcPs
+withInstanceContext templateName tyVars =
+  let tInstanceClass = mkUnqualClass $ templateName ++ "Instance"
```
We should have one single place where we do this `++ "Instance"` thing. We might want to change it to `++ ":Instance"` or something similar in the future and it would be great if we only need to change one place.

rohanjr replied:
Good point

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2743,28 +2813,31 @@ flexChoiceToCombinedChoice (FlexChoiceData controller choiceData) =
 mkTemplateDecls
-  :: Located RdrName -- The template name
-  -> LHsType GhcPs -- The template's record type
+  :: Located TemplateHeader              -- Template constraints, name and type variables
```
Why not make it proper documentation?
```suggestion
  :: Located TemplateHeader              -- ^ Template constraints, name and type variables
```

rohanjr replied:
Will do

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2743,28 +2813,31 @@ flexChoiceToCombinedChoice (FlexChoiceData controller choiceData) =
-      keyInstanceDecl = mkKeyInstanceDecl templateName <$> kdKeyType . unLoc <$> vtbKeyData
+      templateInstClassDecl = mkTemplateInstanceClassDecl nloc conName th vtb{vtbChoices = choicesWithArchive}
+      -- Automatically create the base class (`TInstance`) instance if the template is not generic (i.e. has no type parameters)
+      baseInstance = if null tyVars then [instDecl $ classInstDecl (unLoc tInstanceClass) emptyBag] else []
```
Nice!

hurryabit reviewed (commented):
Very nice! I have a few ideas how to simplify it bit. Let's hope it works.

Thanks a lot! 🙇

rohanjr reviewed (commented):
Thanks for the comments! Left some follow-ups, and will do a fix-up pass

rohanjr committed:
Parse template instance arguments as types instead of qtycons

rohanjr committed:
Use ghc helper function

rohanjr committed:
Remove redundant parse rule

rohanjr committed:
Use haddock syntax and encapsulate TInstance naming scheme

rohanjr committed:
Parse template application as a single type application and split it in mkTemplateInstance

rohanjr commented:
@martin-drhu-da I've addressed your comments that we decided to fix now. https://github.com/digital-asset/ghc/pull/32/commits/31436f5d534e13a36c8e3a6d1e09801d1d3a8ff6 fixes a bug in parsing type arguments in template instances, you may want to have a look before I merge this. I've tested in Haskell and will test again on the DAML side.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2781,6 +2858,35 @@ mkTemplateDecls lname@(L nloc name) fields (L _ decls) = do
+      newTypeDecl <- mkTyData instLoc NewType Nothing (noLoc (Nothing, rdrNameToType instName)) Nothing [newTypeCon] (noLoc [])
+      return $ toOL [TyClD noExt <$> newTypeDecl, instDecl $ classInstDecl instType emptyBag]
+  | otherwise = addFatalError instLoc $ text $
+      occNameString (rdrNameOcc (unLoc instName)) ++ " is not an application of a generic template"
```
This should make your TODO obsolete:
```suggestion
      getOccString instName ++ " is not an application of a generic template"
```

rohanjr replied:
`getOccString` doesn't work because there is no `NamedThing` instance for `RdrName` - it seems like there should be but maybe it's intentionally omitted for some reason. Anyway I'm adding a helper function for now.

hurryabit commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2781,6 +2858,35 @@ mkTemplateDecls lname@(L nloc name) fields (L _ decls) = do
+mkTemplateInstance instName@(L instLoc _) templateApp
+  | (templateType, tyArgs) <- splitHsAppTysPs templateApp
+  , L _ (HsTyVar NoExt NotPromoted templateName) <- templateType = do
+      let templateString = occNameString $ rdrNameOcc $ unLoc templateName
```
```suggestion
      let templateString = getOccString templateName
```
This applies below as well since it works with located and not located stuff. There's also `getOccFS` which is faster and plays well with, say, `mkDataOccFS`.

hurryabit reviewed (approved):
Even better. Feel free to address the `getOccName` and `getOccFS` comments in a follow-up PR.

rohanjr committed:
Use rdrNameToString and rdrNameToType

rohanjr committed:
Simplify templateConstraintToType
